File: projects/inflection/modules/dev/dev_py/ru/declension/data/endings/adj.py
from projects.inflection.modules.dev.dev_py import a
from projects.inflection.modules.dev.dev_py import mw
from projects.inflection.modules.dev.dev_py import tools as _

dev_prefix = 'User:Vitalik/'  # comment this on `prod` version

import sys, os; sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
try:
    from data.endings.noun import copy_other_endings, dump_data
except ImportError:
    from ...data.endings.noun import copy_other_endings, dump_data


# constants:
unstressed = 0  # local
stressed = 1  # local
genders = ['m', 'n', 'f', 'pl']  # local

module = 'data.endings.adj'  # local


# Данные: все стандартные окончания для двух типов основ
@a.call(module)
def base_adj_endings():
    return {
        'm': {
            '1-hard': {
                'nom-sg': ['ый', 'ой'],
                'gen-sg': ['ого', 'ого'],
                'dat-sg': ['ому', 'ому'],
                'acc-sg': '...',
                'ins-sg': 'ым',
                'prp-sg': ['ом', 'ом'],
                'srt-sg': '',
            },
            '2-soft': {
                'nom-sg': 'ий',
                'gen-sg': 'его',
                'dat-sg': 'ему',
                'acc-sg': '...',
                'ins-sg': 'им',
                'prp-sg': 'ем',
                'srt-sg': 'ь',
            },
        },
        'f': {
            '1-hard': {
                'nom-sg': 'ая',
                'gen-sg': ['ой', 'ой'],
                'dat-sg': ['ой', 'ой'],
                'acc-sg': 'ую',
                'ins-sg': ['ой', 'ой'],
                'prp-sg': ['ой', 'ой'],
                'srt-sg': 'а',
            },
            '2-soft': {
                'nom-sg': 'яя',
                'gen-sg': 'ей',
                'dat-sg': 'ей',
                'acc-sg': 'юю',
                'ins-sg': 'ей',
                'prp-sg': 'ей',
                'srt-sg': 'я',
            },
        },
        'n': {
            '1-hard': {
                'nom-sg': ['ое', 'ое'],
                'gen-sg': ['ого', 'ого'],
                'dat-sg': ['ому', 'ому'],
                'acc-sg': '...',
                'ins-sg': 'ым',
                'prp-sg': ['ом', 'ом'],
                'srt-sg': ['о', 'о'],
            },
            '2-soft': {
                'nom-sg': 'ее',
                'gen-sg': 'его',
                'dat-sg': 'ему',
                'acc-sg': '...',
                'ins-sg': 'им',
                'prp-sg': 'ем',
                'srt-sg': ['е', 'ё'],
            },
        },
        'pl': {  # plural endings
            '1-hard': {
                'nom-pl': 'ые',
                'gen-pl': 'ых',
                'dat-pl': 'ым',
                'acc-pl': '...',
                'ins-pl': 'ыми',
                'prp-pl': 'ых',
                'srt-pl': 'ы',
            },
            '2-soft': {
                'nom-pl': 'ие',
                'gen-pl': 'их',
                'dat-pl': 'им',
                'acc-pl': '...',
                'ins-pl': 'ими',
                'prp-pl': 'их',
                'srt-pl': 'и',
            },
        },
    }
# end


# Изменение окончаний для остальных типов основ (базирующихся на первых двух)
@a.starts(module)
def fill_other_adj_endings(func, endings):
    # INFO: Replace "ы" to "и"
    for j, stem_type in enumerate(['3-velar', '4-sibilant']):
        endings['m'][stem_type]['nom-sg'][unstressed] = 'ий'  # adj only
        endings['m'][stem_type]['ins-sg'] = 'им'
        endings['n'][stem_type]['ins-sg'] = 'им'

        endings['pl'][stem_type]['nom-pl'] = 'ие'  # adj only
        # endings['pl'][stem_type]['nom-pl'] = 'и'  # pronoun only
        endings['pl'][stem_type]['gen-pl'] = 'их'
        endings['pl'][stem_type]['dat-pl'] = 'им'
        endings['pl'][stem_type]['ins-pl'] = 'ими'
        endings['pl'][stem_type]['prp-pl'] = 'их'
        endings['pl'][stem_type]['srt-pl'] = 'и'  # adj only
    # end

    # INFO: Replace unstressed "о" to "е"
    for j, stem_type in enumerate(['4-sibilant', '5-letter-ц']):
        # if not i.stress_schema['ending']['sg']:  # fixme
        endings['m'][stem_type]['nom-sg'][stressed] = 'ей'  # adj only  # fixme ???
        endings['m'][stem_type]['gen-sg'][unstressed] = 'его'
        endings['m'][stem_type]['dat-sg'][unstressed] = 'ему'
        endings['m'][stem_type]['prp-sg'][unstressed] = 'ем'

        endings['n'][stem_type]['nom-sg'][unstressed] = 'ее'
        endings['n'][stem_type]['gen-sg'][unstressed] = 'его'
        endings['n'][stem_type]['dat-sg'][unstressed] = 'ему'
        endings['n'][stem_type]['prp-sg'][unstressed] = 'ем'

        endings['f'][stem_type]['gen-sg'][unstressed] = 'ей'
        endings['f'][stem_type]['dat-sg'][unstressed] = 'ей'
        endings['f'][stem_type]['ins-sg'][unstressed] = 'ей'
        endings['f'][stem_type]['prp-sg'][unstressed] = 'ей'

        endings['n'][stem_type]['srt-sg'][unstressed] = 'е'  # adj only
    # end

    # INFO: Replace "ь" to "й"
    stem_type = '6-vowel'  # local
    endings['m'][stem_type]['srt-sg'] = 'й'  # adj only

    _.ends(module, func)
# end


@a.starts(module)
def generate_adj_endings(func):
    endings = base_adj_endings()  # local
    copy_other_endings(endings, genders)
    fill_other_adj_endings(endings)
    dump_data('adj', endings)
    return _.returns(module, func, endings)
# end


if __name__ == '__main__':
    generate_adj_endings()

File: projects/inflection/modules/dev/dev_lua/ru/declension/data/endings/adj.lua
local dev_prefix = ''
dev_prefix = 'User:Vitalik/'  -- comment this on `prod` version

local export = {}
local _ = require('Module:' .. dev_prefix .. 'inflection/tools')


-- constants:
local unstressed = 1
local stressed = 2
local module = 'data.endings.adj'


-- Данные: все стандартные окончания для двух типов основ
-- @call
function export.get_standard_adj_endings()
	func = "get_standard_adj_endings"
	_.call(module, func)

	local e = {}  -- AttrDict
	e.m = {}  -- AttrDict
	e.f = {}  -- AttrDict
	e.n = {}  -- AttrDict
	e.common = {}  -- AttrDict
	e.m.hard = {}  -- dict
	e.m.soft = {}  -- dict
	e.f.hard = {}  -- dict
	e.f.soft = {}  -- dict
	e.n.hard = {}  -- dict
	e.n.soft = {}  -- dict
	e.common.hard = {}  -- dict
	e.common.soft = {}  -- dict
	e.m.hard['nom-sg'] = {'ый', 'ой'}
	e.m.hard['gen-sg'] = 'ого'
	e.m.hard['dat-sg'] = 'ому'
	e.m.hard['ins-sg'] = 'ым'
	e.m.hard['prp-sg'] = 'ом'
	e.m.hard['srt-sg'] = ''
	e.m.soft['nom-sg'] = 'ий'
	e.m.soft['gen-sg'] = 'его'
	e.m.soft['dat-sg'] = 'ему'
	e.m.soft['ins-sg'] = 'им'
	e.m.soft['prp-sg'] = 'ем'
	e.m.soft['srt-sg'] = 'ь'
	e.f.hard['nom-sg'] = 'ая'
	e.f.hard['gen-sg'] = 'ой'
	e.f.hard['dat-sg'] = 'ой'
	e.f.hard['acc-sg'] = 'ую'
	e.f.hard['ins-sg'] = 'ой'
	e.f.hard['prp-sg'] = 'ой'
	e.f.hard['srt-sg'] = 'а'
	e.f.soft['nom-sg'] = 'яя'
	e.f.soft['gen-sg'] = 'ей'
	e.f.soft['dat-sg'] = 'ей'
	e.f.soft['acc-sg'] = 'юю'
	e.f.soft['ins-sg'] = 'ей'
	e.f.soft['prp-sg'] = 'ей'
	e.f.soft['srt-sg'] = 'я'
	e.n.hard['nom-sg'] = 'ое'
	e.n.hard['gen-sg'] = 'ого'
	e.n.hard['dat-sg'] = 'ому'
	e.n.hard['ins-sg'] = 'ым'
	e.n.hard['prp-sg'] = 'ом'
	e.n.hard['srt-sg'] = 'о'
	e.n.soft['nom-sg'] = 'ее'
	e.n.soft['gen-sg'] = 'его'
	e.n.soft['dat-sg'] = 'ему'
	e.n.soft['ins-sg'] = 'им'
	e.n.soft['prp-sg'] = 'ем'
	e.n.soft['srt-sg'] = {'е', 'ё'}
	e.common.hard['nom-pl'] = 'ые'
	e.common.hard['gen-pl'] = 'ых'
	e.common.hard['dat-pl'] = 'ым'
	e.common.hard['ins-pl'] = 'ыми'
	e.common.hard['prp-pl'] = 'ых'
	e.common.hard['srt-pl'] = 'ы'
	e.common.soft['nom-pl'] = 'ие'
	e.common.soft['gen-pl'] = 'их'
	e.common.soft['dat-pl'] = 'им'
	e.common.soft['ins-pl'] = 'ими'
	e.common.soft['prp-pl'] = 'их'
	e.common.soft['srt-pl'] = 'и'
	return e
end


-- Изменение окончаний для остальных типов основ (базирующихся на первых двух)
-- @starts
function export.fix_adj_pronoun_endings(endings, gender, stem_type, stress_schema, adj, pronoun)
	func = "fix_adj_pronoun_endings"
	_.starts(module, func)

--	INFO: Replace "ы" to "и"
	if _.equals(stem_type, {'velar', 'sibilant'}) then
		if gender == 'm' then
			if adj then
				endings['nom-sg'][unstressed] = 'ий'
			end
			endings['ins-sg'] = 'им'
		end
		if gender == 'n' then
			endings['ins-sg'] = 'им'
		end

		if adj then
			endings['nom-pl'] = 'ие'
		elseif pronoun then
			endings['nom-pl'] = 'и'
		end
		endings['gen-pl'] = 'их'
		endings['dat-pl'] = 'им'
		endings['ins-pl'] = 'ими'
		endings['prp-pl'] = 'их'
		if adj then
			endings['srt-pl'] = 'и'
		end
	end

--	INFO: Replace unstressed "о" to "е"
	if _.equals(stem_type, {'sibilant', 'letter-ц'}) then
		if not stress_schema['ending']['sg'] then
			if gender == 'm' then
				if adj then
					endings['nom-sg'][stressed] = 'ей'
				end
				endings['gen-sg'] = 'его'
				endings['dat-sg'] = 'ему'
				endings['prp-sg'] = 'ем'
			end
			if gender == 'n' then
				endings['nom-sg'] = 'ее'
				endings['gen-sg'] = 'его'
				endings['dat-sg'] = 'ему'
				endings['prp-sg'] = 'ем'
			end
			if gender == 'f' then
				endings['gen-sg'] = 'ей'
				endings['dat-sg'] = 'ей'
				endings['ins-sg'] = 'ей'
				endings['prp-sg'] = 'ей'
			end
		end
		if not stress_schema['ending']['srt-sg-n'] then
			if gender == 'n' then
				if adj then
					endings['srt-sg'] = 'е'
				end
			end
		end
	end

--	INFO: Replace "ь" to "й"
	if _.equals(stem_type, {'vowel'}) then
		if gender == 'm' then
			if adj then
				endings['srt-sg'] = 'й'
			end
		end
	end

	_.ends(module, func)
end


return export
